chalice/deploy/packager.py

```diff
@@ -208,3 +216,373 @@ def _add_chalice_lib_if_needed(self, project_dir, zip_fileobj):
+        # are present.
+        data_dir = self._osutils.joinpath(root, whl.data_dir)
+        if not self._osutils.directory_exists(data_dir):
+            return
+        unpack_dirs = {'purelib', 'platlib'}
+        data_contents = self._osutils.get_directory_contents(data_dir)
+        for content_name in data_contents:
+            if content_name in unpack_dirs:
+                source = self._osutils.joinpath(data_dir, content_name)
+                self._osutils.copytree(source, root)
+                # No reason to keep the purelib/platlib source directory around
+                # so we delete it to conserve space in the package.
+                self._osutils.rmtree(source)
+
+    def _install_whls(self, src_dir, dst_dir, whls):
+        # type: (str, str, Set[Package]) -> None
+        if os.path.isdir(dst_dir):
+            shutil.rmtree(dst_dir)
+        os.makedirs(dst_dir)
+        for whl in whls:
+            zipfile_path = self._osutils.joinpath(src_dir, whl.filename)
+            with zipfile.ZipFile(zipfile_path, 'r') as z:
+                z.extractall(dst_dir)
+            self._install_purelib_and_platlib(whl, dst_dir)
+
+    def build_site_packages(self, project_dir):
+        # type: (str) -> None
+        requirements_file = self._osutils.joinpath(
+            project_dir, 'requirements.txt')
+        deps_dir = self.site_package_dir(project_dir)
+        if self._has_at_least_one_package(requirements_file):
+            with self._osutils.tempdir() as tempdir:
+                valid_whls, missing_whls = self._download_dependencies(
+                    tempdir, requirements_file)
+                self._install_whls(tempdir, deps_dir, valid_whls)
+            if missing_whls:
+                raise MissingDependencyError(missing_whls)
+
+    def site_package_dir(self, project_dir):
+        # type: (str) -> str
+        """Return path to the site packages directory."""
+        deps_dir = self._osutils.joinpath(
+            project_dir, '.chalice', 'site-packages')
+        return deps_dir
+
+
+class Package(object):
+    PYPI_SDIST_EXTS = ['.zip', '.tar.gz']
+
+    def __init__(self, directory, filename):
+        # type: (str, str) -> None
+        self.dist_type = 'whl' if filename.endswith('whl') else 'sdist'
+        self.filename = filename
+        self._directory = directory
+        self._name, self._version = self._calculate_name_and_version()
+
+    @property
+    def data_dir(self):
+        # type: () -> str
+        # The directory format is {distribution}-{version}.data
+        return '%s-%s.data' % (self._name, self._version)
+
+    @property
+    def identifier(self):
+        # type: () -> str
+        return '%s==%s' % (self._name, self._version)
+
+    def __str__(self):
+        # type: () -> str
+        return '%s(%s)' % (self.identifier, self.dist_type)
+
+    def __repr__(self):
+        # type: () -> str
+        return str(self)
+
+    def __eq__(self, other):
```

Comment:
Not equal doesn't really make sense in this context and the equal set and unequal set are not disjoint so I don't think its worth implementing as it should not be used and currently isn't.